R/zzz.R

```diff
@@ -151,6 +151,20 @@ pl$mem_address = mem_address
     env = pl
   )
 
+  makeActiveBinding(
+    "rpool_avail",
+    function() {
+      unwrap(get_global_rpool_cap())[["available"]]
+    },
+    env = polars_optenv
+  )
+
+  rpool_cap = unwrap(get_global_rpool_cap())[["capacity"]]
+  polars_optenv$rpool_cap = rpool_cap
+  # the max number of R session is the number returned the first time, this is
+  # not updated after because it will always be the limit
+  polars_optenv$rpool_cap_max = rpool_cap
```

Comment:
I don't there should be such a limit as rpool_cap_max. The user can set cap to whatever non negative integer at any time. "Available" is the current number of spawned processes in pool. cap is the total number allowed. If cap is lowered below available, while processes are in the idle queue, they are just shut down. If cap is lowered while processes are running, any running process is allowed to finish it's current job gracefully, before shutting down. If a new job, do not find an idle process ind the "idle process queue", it will spawn a new processes and increment the available counter. However if that would exceed the current cap, the job-thread will "go to sleep" in the "job-queue" and wait for being picked up by a idling process.

We can advice the user that setting cap more than n-cores is probably not faster. Sometimes in benchmarks 150% processes more than cores did slightly better. Maybe someone has a constant IO limit (e.g. scraping) and several more processes is actually faster.

Polars has by default 2xnCPU threads, so exceeding would likely not do anything because all threads are saturated with idle processes. Last I checked the rust-polars thread cap is actually static and can only be modified with an envvar at startup. But rpool_cap does not have such a limitation.

Reply:
current cap logic
https://github.com/pola-rs/r-polars/blob/68e4911a8e20b0e2d80fbc0c49ca48b6bab740e9/src/rust/src/rbackground.rs#L321

Reply:
> The user can set cap to whatever non negative integer at any time

But the original value for capacity is the value returned by `RBGPOOL` so can it really be set above this value? If so I can remove this requirement and add a message if the user sets the capacity above the default of `RBGPOOL`

>  "Available" is the current number of spawned processes in pool. cap is the total number allowed in the very moment

By the way, "available" is a strange name for this, shouldn't it be "active" instead since it returns the number of processes running in the background? Also, in `get_global_rpool_cap()` in the Rust side, the result of `RBGPOOL` is named "active":
```rust
pub fn get_global_rpool_cap() -> RResult<List> {
    let pool_guard = RBGPOOL.0.lock()?;
    Ok(list!(
        available = pool_guard.active,
        capacity = pool_guard.cap
    ))
}
```

Reply:
RBGPool has type static once_cell::sync::lazy<RBackGroundPool> This means it is a lazy stacsingleton which there is only one global instance of, and it comes into existence lazily when accessed the first time.

RBackGroundPool is a big state, which contains the process pool and the job pool. Any Polars worker can one at the time acquire the Mutex lock. Then the polars worker can try "lease" a process or sleep in the job queue until a process is availble. The interaction with the RBGPool is very fast as it just book keeping, so the thread will not block each other. When a polars worker is done with a process, it will ask RBGPool what other thread to pass on the R proccess to.

```R
Restarting R session...

> library(polars)
>   pl$get_global_rpool_cap()
$available
[1] 0

$capacity
[1] 4

>   pl$set_global_rpool_cap(5)
[src/rbackground.rs:440] "resize" = "resize"
[src/rbackground.rs:440] &self = RBackgroundPool(
    Mutex {
        data: InnerRBackgroundPool {
            pool: [],
            queue: [],
            active: 0,
            cap: 4,
        },
        poisoned: false,
        ..
    },
)
>   pl$set_global_rpool_cap(5)
[src/rbackground.rs:440] "resize" = "resize"
[src/rbackground.rs:440] &self = RBackgroundPool(
    Mutex {
        data: InnerRBackgroundPool {
            pool: [],
            queue: [],
            active: 0,
            cap: 5,
        },
        poisoned: false,
        ..
    },
)
>   l_expr = lapply(1:11,\(i) pl$lit(i)$map(\(x) {Sys.sleep(i/5); -i}, in_background = TRUE)$alias(paste0("lit_",i)))
>   lf = pl$LazyFrame()$select(l_expr)
>   unlist(pl$get_global_rpool_cap()) |> (\(x) paste(names(x),unname(x)))() |> print()
[1] "available 0" "capacity 5" 
>   handle = lf$collect(collect_in_background = TRUE)
>   pl$set_global_rpool_cap(3)
[src/rbackground.rs:440] "resize" = "resize"
[src/rbackground.rs:440] &self = RBackgroundPool(
    Mutex {
        data: InnerRBackgroundPool {
            pool: [],
            queue: [],
            active: 0,
            cap: 5,
        },
        poisoned: false,
        ..
    },
)
>   for (i in 1:10) {
+     unlist(pl$get_global_rpool_cap()) |> (\(x) paste(names(x),unname(x)))() |> print()
[src/rbackground.rs:477] "rmap_series" = "rmap_series"
+     Sys.sleep(.3)
lease a newly created handler
lease a newly created handler
lease a newly created handler
[src/rbackground.rs:385] "METHOD lease" = "METHOD lease"
[src/rbackground.rs:477] "rmap_series" = "rmap_series"
[src/rbackground.rs:385] &self = RBackgroundPool(
    Mutex {
        data: InnerRBackgroundPool {
            pool: [],
            queue: [],
            active: 0,
            cap: 3,
        },
        poisoned: false,
        ..
    },
)
[src/rbackground.rs:477] "rmap_series" = "rmap_series"
[src/rbackground.rs:385] "METHOD lease" = "METHOD lease"
[src/rbackground.rs:385] &self = RBackgroundPool(
    Mutex {
        data: <locked>,
        poisoned: false,
        ..
    },
)
[src/rbackground.rs:477] "rmap_series" = "rmap_series"
[src/rbackground.rs:385] "METHOD lease" = "METHOD lease"
[src/rbackground.rs:477] "rmap_series" = "rmap_series"
[src/rbackground.rs:385] "METHOD lease" = "METHOD lease"
[src/rbackground.rs:385] &self = RBackgroundPool(
    Mutex {
        data: InnerRBackgroundPool {
            pool: [],
            queue: [],
            active: 2,
            cap: 3,
        },
        poisoned: false,
        ..
    },
)
active == cap, no more handlers allowed, go into queue
wait for freed handler
[src/rbackground.rs:385] "METHOD lease" = "METHOD lease"
[src/rbackground.rs:385] &self = RBackgroundPool(
    Mutex {
        data: InnerRBackgroundPool {
            pool: [],
            queue: [],
            active: 3,
            cap: 3,
        },
        poisoned: false,
        ..
    },
)
[src/rbackground.rs:385] &self = RBackgroundPool(
    Mutex+   }
active == cap, no more handlers allowed, go into queue
wait for freed handler
 {
        data: InnerRBackgroundPool {
            pool: [],
            queue: [
                Sender,
            ],
            active: 3,
            cap: 3,
        },
        poisoned: false,
        ..
    },
)
[src/rbackground.rs:477] "rmap_series" = "rmap_series"
[src/rbackground.rs:385] "METHOD lease" = "METHOD lease"
[src/rbackground.rs:385] &self = RBackgroundPool(
    Mutex {
        data: InnerRBackgroundPool {
            pool: [],
            queue: [
                Sender,
                Sender,
            ],
            active: 3,
            cap: 3,
        },
        poisoned: false,
        ..
    },
)
active == cap, no more handlers allowed, go into queue
wait for freed handler
[src/rbackground.rs:477] "rmap_series" = "rmap_series"
[src/rbackground.rs:385] "METHOD lease" = "METHOD lease"
[src/rbackground.rs:385] &self = RBackgroundPool(
    Mutex {
        data: InnerRBackgroundPool {
            pool: [],
            queue: [
                Sender,
                Sender,
                Sender,
            ],
            active: 3,
            cap: 3,
        },
        poisoned: false,
        ..
    },
)
active == cap, no more handlers allowed, go into queue
wait for freed handler
[src/rbackground.rs:477] "rmap_series" = "rmap_series"
[src/rbackground.rs:385] "METHOD lease" = "METHOD lease"
[src/rbackground.rs:385] &self = RBackgroundPool(
    Mutex {
        data: InnerRBackgroundPool {
            pool: [],
            queue: [
                Sender,
                Sender,
                Sender,
                Sender,
            ],
            active: 3,
            cap: 3,
        },
        poisoned: false,
        ..
    },
)
active == cap, no more handlers allowed, go into queue
wait for freed handler
[1] "available 3" "capacity 3" 
[1] "available 3" "capacity 3" 
[1] "available 3" "capacity 3" 
[src/rbackground.rs:431] "shelf" = "shelf"
[src/rbackground.rs:431] &self = RBackgroundPool(
    Mutex {
        data: InnerRBackgroundPool {
            pool: [],
            queue: [
                Sender,
                Sender,
                Sender,
                Sender,
                Sender,
            ],
            active: 3,
            cap: 3,
        },
        poisoned: false,
        ..
    },
)
2a - fine, pass it on!
thread was awoken queue and passed a handler
[src/rbackground.rs:477] "rmap_series" = "rmap_series"
[src/rbackground.rs:385] "METHOD lease" = "METHOD lease"
[src/rbackground.rs:385] &self = RBackgroundPool(
    Mutex {
        data: InnerRBackgroundPool {
            pool: [],
            queue: [
                Sender,
                Sender,
                Sender,
                Sender,
            ],
            active: 3,
            cap: 3,
        },
        poisoned: false,
        ..
    },
)
active == cap, no more handlers allowed, go into queue
wait for freed handler
[1] "available 3" "capacity 3" 
[1] "available 3" "capacity 3" 
[src/rbackground.rs:431] "shelf" = "shelf"
[src/rbackground.rs:431] &self = RBackgroundPool(
    Mutex {
        data: InnerRBackgroundPool {
            pool: [],
            queue: [
                Sender,
                Sender,
                Sender,
                Sender,
                Sender,
            ],
            active: 3,
            cap: 3,
        },
        poisoned: false,
        ..
    },
)
2a - fine, pass it on!
thread was awoken queue and passed a handler
[src/rbackground.rs:477] "rmap_series" = "rmap_series"
[src/rbackground.rs:385] "METHOD lease" = "METHOD lease"
[src/rbackground.rs:385] &self = RBackgroundPool(
    Mutex {
        data: InnerRBackgroundPool {
            pool: [],
            queue: [
                Sender,
                Sender,
                Sender,
                Sender,
            ],
            active: 3,
            cap: 3,
        },
        poisoned: false,
        ..
    },
)
active == cap, no more handlers allowed, go into queue
wait for freed handler
[src/rbackground.rs:431] "shelf" = "shelf"
[src/rbackground.rs:431] &self = RBackgroundPool(
    Mutex {
        data: InnerRBackgroundPool {
            pool: [],
            queue: [
                Sender,
                Sender,
                Sender,
                Sender,
                Sender,
            ],
            active: 3,
            cap: 3,
        },
        poisoned: false,
        ..
    },
)
[src/rbackground.rs:431] "shelf" = "shelf"
[src/rbackground.rs:431] &self = RBackgroundPool(
    Mutex {
        data: InnerRBackgroundPool {
            pool: [],
            queue: [
                Sender,
                Sender,
                Sender,
                Sender,
            ],
            active: 3,
            cap: 3,
        },
        poisoned: false,
        ..
    },
)
2a - fine, pass it on!
thread was awoken queue and passed a handler
2a - fine, pass it on!
thread was awoken queue and passed a handler
[src/rbackground.rs:477] "rmap_series" = "rmap_series"
[src/rbackground.rs:385] "METHOD lease" = "METHOD lease"
[src/rbackground.rs:385] &self = RBackgroundPool(
    Mutex {
        data: InnerRBackgroundPool {
            pool: [],
            queue: [
                Sender,
                Sender,
                Sender,
            ],
            active: 3,
            cap: 3,
        },
        poisoned: false,
        ..
    },
)
active == cap, no more handlers allowed, go into queue
wait for freed handler
[1] "available 3" "capacity 3" 
[src/rbackground.rs:431] "shelf" = "shelf"
[src/rbackground.rs:431] &self = RBackgroundPool(
    Mutex {
        data: InnerRBackgroundPool {
            pool: [],
            queue: [
                Sender,
                Sender,
                Sender,
                Sender,
            ],
            active: 3,
            cap: 3,
        },
        poisoned: false,
        ..
    },
)
2a - fine, pass it on!
thread was awoken queue and passed a handler
[1] "available 3" "capacity 3" 
[1] "available 3" "capacity 3" 
[1] "available 3" "capacity 3" 
[src/rbackground.rs:431] "shelf" = "shelf"
[src/rbackground.rs:431] &self = RBackgroundPool(
    Mutex {
        data: InnerRBackgroundPool {
            pool: [],
            queue: [
                Sender,
                Sender,
                Sender,
            ],
            active: 3,
            cap: 3,
        },
        poisoned: false,
        ..
    },
)
2a - fine, pass it on!
thread was awoken queue and passed a handler
[1] "available 3" "capacity 3" 
>   df = handle$join()
[src/rbackground.rs:431] "shelf" = "shelf"
[src/rbackground.rs:431] &self = RBackgroundPool(
    Mutex {
        data: InnerRBackgroundPool {
            pool: [],
            queue: [
                Sender,
                Sender,
            ],
            active: 3,
            cap: 3,
        },
        poisoned: false,
        ..
    },
)
2a - fine, pass it on!
thread was awoken queue and passed a handler
[src/rbackground.rs:431] "shelf" = "shelf"
[src/rbackground.rs:431] &self = RBackgroundPool(
    Mutex {
        data: InnerRBackgroundPool {
            pool: [],
            queue: [
                Sender,
            ],
            active: 3,
            cap: 3,
        },
        poisoned: false,
        ..
    },
)
2a - fine, pass it on!
thread was awoken queue and passed a handler
[src/rbackground.rs:431] "shelf" = "shelf"
[src/rbackground.rs:431] &self = RBackgroundPool(
    Mutex {
        data: InnerRBackgroundPool {
            pool: [],
            queue: [],
            active: 3,
            cap: 3,
        },
        poisoned: false,
        ..
    },
)
3a - park handler in pool
[src/rbackground.rs:431] "shelf" = "shelf"
[src/rbackground.rs:431] &self = RBackgroundPool(
    Mutex {
        data: InnerRBackgroundPool {
            pool: [
                RBackgroundHandler {
                    proc: Child {
                        stdin: None,
                        stdout: None,
                        stderr: None,
                        ..
                    },
                    chan: IpcSender {
                        os_sender: OsIpcSender {
                            port: 22019,
                            nosync_marker: PhantomData<core::cell::Cell<()>>,
                        },
                        phantom: PhantomData<r_polars::rbackground::RIPCJob>,
                    },
                    in_pool: true,
                },
            ],
            queue: [],
            active: 3,
            cap: 3,
        },
        poisoned: false,
        ..
    },
)
3a - park handler in pool
[src/rbackground.rs:431] "shelf" = "shelf"
[src/rbackground.rs:431] &self = RBackgroundPool(
    Mutex {
        data: InnerRBackgroundPool {
            pool: [
                RBackgroundHandler {
                    proc: Child {
                        stdin: None,
                        stdout: None,
                        stderr: None,
                        ..
                    },
                    chan: IpcSender {
                        os_sender: OsIpcSender {
                            port: 22019,
                            nosync_marker: PhantomData<core::cell::Cell<()>>,
                        },
                        phantom: PhantomData<r_polars::rbackground::RIPCJob>,
                    },
                    in_pool: true,
                },
                RBackgroundHandler {
                    proc: Child {
                        stdin: None,
                        stdout: None,
                        stderr: None,
                        ..
                    },
                    chan: IpcSender {
                        os_sender: OsIpcSender {
                            port: 15875,
                            nosync_marker: PhantomData<core::cell::Cell<()>>,
                        },
                        phantom: PhantomData<r_polars::rbackground::RIPCJob>,
                    },
                    in_pool: true,
                },
            ],
            queue: [],
            active: 3,
            cap: 3,
        },
        poisoned: false,
        ..
    },
)
3a - park handler in pool
>   print(df)
shape: (1, 11)
┌───────┬───────┬───────┬───────┬───┬───────┬───────┬────────┬────────┐
│ lit_1 ┆ lit_2 ┆ lit_3 ┆ lit_4 ┆ … ┆ lit_8 ┆ lit_9 ┆ lit_10 ┆ lit_11 │
│ ---   ┆ ---   ┆ ---   ┆ ---   ┆   ┆ ---   ┆ ---   ┆ ---    ┆ ---    │
│ i32   ┆ i32   ┆ i32   ┆ i32   ┆   ┆ i32   ┆ i32   ┆ i32    ┆ i32    │
╞═══════╪═══════╪═══════╪═══════╪═══╪═══════╪═══════╪════════╪════════╡
│ -1    ┆ -2    ┆ -3    ┆ -4    ┆ … ┆ -8    ┆ -9    ┆ -10    ┆ -11    │
└───────┴───────┴───────┴───────┴───┴───────┴───────┴────────┴────────┘
>   42
[1] 42
```

However RBackGroundPool is very mutable it has an outer Mutex which and thread can acces

Reply:
Thanks for the explanations, I removed the constraint of `cap_max` and added a message, are you ok with this?

Also, can I replace "available" by "active" when talking about the number of processes running in the background? (cf my previous comment)

Reply:
I have tried to remove cap_max now and fiddled with errors and set_rpool_cap an active binding also.

I think in hindsight a more suitable code pattern might have been, that every option just had a getter and a setter which check args if needed. Then pl$set_options would just iterate over setters and pl$options over the getters. But as is now is pretty fine also.